src/main/java/com/ebivariation/contigalias/dao/FTPClientAssemblyDaoImplement.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2020 EMBL - European Bioinformatics Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ebivariation.contigalias.dao;
+
+import com.ebivariation.contigalias.dus.AssemblyReportReader;
+import com.ebivariation.contigalias.dus.NCBIBrowser;
+import com.ebivariation.contigalias.entities.AssemblyEntity;
+import org.springframework.stereotype.Repository;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+@Repository("ftpDao")
+public class FTPClientAssemblyDaoImplement implements AssemblyDao {
+
+    @Override
+    public AssemblyEntity getAssemblyByAccession(String accession) throws IOException {
+        NCBIBrowser ncbiBrowser = new NCBIBrowser();
+        ncbiBrowser.connect();
+        String directory = ncbiBrowser.getGenomeReportDirectory(accession);
+        String path = ncbiBrowser.getPathFromRoot(directory);
+        InputStream stream = ncbiBrowser.getAssemblyReportInputStream(path);
+        InputStreamReader streamReader = new InputStreamReader(stream);
+        AssemblyReportReader reader = new AssemblyReportReader(streamReader);
```

Comment:
Added a constructor to accept an InputStream. Would it be a good idea to have AssemblyReportReader also extend BufferedReader instead of having it as a local object?

Reply:
Relevant commit with required changes https://github.com/singaltanmay/contig-alias/commit/6c9f15ab9ab9ade2c08a73103150368cbd9d0d99

Reply:
In general you should favor composition over inheritance. In the case of the FTPBrowser I was on the fence but I kind of agreed it was better to extend. In this case I think it's not a good idea. Please revert 6c9f15a. If you are interested I can explain in more detail why I think like this, both for this case and for the general case.

Reply:
what I actually missed is that we are not closing the reader. We should close it after parsing.

Independently to that, if you want your AssemblyReportReader to be more like a real Reader, what you should do is `implements Reader` instead of extending.

Reply:
`6c9f15a` was on a different branch and I have deleted that branch. I would like to know more about when to extends and when to not.
I have closed the BufferedReader in the last line of parseReport(). 
Does closing the BufferedReader automatically the InputStreamReader provided to it as well? And does the InputStreamReader then close the InputStream? And what about closing stuff in the other direction? Will closing the stream close all readers recursively?
Implementing a custom Reader means writing our own read() and close(). That seems like too much low level work with no real advantage in our case.

Reply:
for the FTPBrowser in #10 you said:
> inheritance is better OOP practice.

Actually, in general, composition is preferred over inheritance. https://en.wikipedia.org/wiki/Composition_over_inheritance . There are several reasons:
- with inheritance, when the class starts to get a bit complicated it's hard to extract functionality and you start getting problems with multiple inheritance.
- OOP shines best for encapsulation and polymorphism. If your concern is none of that, and you only worry about reusing functionality, inheritance only removes some boilerplate code. And actually, if reusing code is your only concern, you would get better results with regular procedural code (static methods in java) instead of OOP.
- Composition fits better with the open-closed principle. This is relevant for adding functionality without modifying the class.

Reply:
if you do ctrl+click in BufferedReader you can see the `close()` implementation, where it closes the inner Reader (the one you provide in the BufferedReader constructor). So BufferedReader knows about Readers and closes them, but Readers are interfaces that don't know about their subclasses (this is the main point of interfaces) so they usually can't close the subclasses.

I agree the AssemblyReportReader shouldn't implement Readable, as it just returns the parsed object, and not chars.